drivers: pwm: Infineon: PWM driver improvements and bug fixes

billwatersiii opened the pull request:
This change makes improvements and bug fixes for the Infineon PWM driver.  These include:
  * Removes hard coded register addresss from driver.
  * Addresses issues causing pwm_api and pwm_gpio_loopback tests to fail, as well as functional failures.
  * Restructures device tree file to better represent the hardware architecture of the tcpwm module.
  * Allows configuration of hardware behavior when PWM is disabled.
  * Changes driver naming to reflect hardware IP being used (TCPWM) instead of referencing cat1.  Cat1 is an internal infineon reference which has little meaning to users and is being phased phased out.
  * Updates the overlay files for samples/basic/blinky_pwm and samples/basic/fade_led to run on the cyw920829m2_evk board.
  * Adds overlay files for tests/drivers/pwm/pwm_api and tests/drivers/pwm/pwm_gpio_loopback to run these tests on the cyw920829 board.

JarmouniA reviewed (commented):
The renaming should happen in a 1st commit, then the actual changes should be split into 3 ones: driver/binding/dts, samples, tests.

billwatersiii committed:
drivers: pwm: Infineon: replace cat1 naming with tcpwm

* Changes driver naming to reflect hardware IP being used (TCPWM)
instead of referencing cat1.  Cat1 is an internal infineon
reference which has little meaning to users and is being phased
phased out.

Signed-off-by: Bill Waters <bill.waters@infineon.com>

billwatersiii committed:
drivers: pwm: Infineon: PWM driver improvements and bug fixes

This change makes improvements and bug fixes for the Infineon PWM
driver.  These include:

* Removes hard coded register addresss from driver.
* Addresses issues causing pwm_api and pwm_gpio_loopback tests to
fail, as well as functional failures.
* Restructures device tree file to better represent the hardware
architecture of the tcpwm module.
* Allows configuration of hardware behavior when PWM is disabled.

Signed-off-by: Bill Waters <bill.waters@infineon.com>

billwatersiii committed:
drivers: pwm: Infineon: updates to PWM driver sample files

* Updates the overlay files for samples/basic/blinky_pwm and
samples/basic/fade_led to run on the cyw920829m2_evk board.

Signed-off-by: Bill Waters <bill.waters@infineon.com>

billwatersiii committed:
drivers: pwm: Infineon: updates to PWM driver test files

* Adds overlay files for tests/drivers/pwm/pwm_api and
tests/drivers/pwm/pwm_gpio_loopback to run these tests on the
cyw920829 board.

Signed-off-by: Bill Waters <bill.waters@infineon.com>

billwatersiii commented:
> The renaming should happen in a 1st commit, then the actual changes should be split into 3 ones: driver/binding/dts, samples, tests.

@JarmouniA , done

JarmouniA reviewed (commented):
I did not comment on all instances.

billwatersiii commented on drivers/pwm/Kconfig.ifx_tcpwm:
```diff
@@ -1,15 +1,14 @@
 
-config PWM_INFINEON_CAT1
-	bool "Infineon CAT1 PWM driver"
+config PWM_INFINEON_TCPWM
```
done

billwatersiii replied:
done

billwatersiii commented on drivers/pwm/Kconfig.ifx_tcpwm:
```diff
@@ -1,15 +1,14 @@
 	default y
-	depends on DT_HAS_INFINEON_CAT1_PWM_ENABLED
-	depends on SOC_FAMILY_INFINEON_CAT1B
+	depends on DT_HAS_INFINEON_TCPWM_PWM_ENABLED
```
done

billwatersiii commented on drivers/pwm/pwm_ifx_tcpwm.c:
```diff
@@ -82,20 +71,21 @@ static int ifx_cat1_pwm_init(const struct device *dev)
 
-static int ifx_cat1_pwm_set_cycles(const struct device *dev, uint32_t channel,
-				   uint32_t period_cycles, uint32_t pulse_cycles, pwm_flags_t flags)
+static int ifx_tcpwm_pwm_set_cycles(const struct device *dev, uint32_t channel,
```
done

billwatersiii commented on drivers/pwm/pwm_ifx_tcpwm.c:
```diff
@@ -108,78 +98,90 @@ static int ifx_cat1_pwm_set_cycles(const struct device *dev, uint32_t channel,
 };
 
-#define INFINEON_CAT1_PWM_INIT(n)                                                                  \
+#define INFINEON_TCPWM_PWM_INIT(n)                                                                 \
```
done

billwatersiii commented on drivers/pwm/pwm_ifx_tcpwm.c:
```diff
@@ -1,50 +1,43 @@
-
-struct ifx_cat1_pwm_config {
-	TCPWM_GRP_CNT_Type *reg_addr;
+struct ifx_tcpwm_pwm_config {
```
done

kartben commented on include/zephyr/drivers/timer/ifx_tcpwm.h:
@billwatersiii it doesn't seem to me as if this should be a public header. Please make it a private header alongside pwm_ifx_tcpwm.c in the drivers/pwm folder if my understanding is correct.

Also, the #include guard is incorrect anyway, it should be fully qualified (`#ifndef ZEPHYR_INCLUDE_DRIVERS_TIMER_IFX_TCPWM_TIMER_H` even if this file were to stay there.

cc @andyross @teburd 

Thanks!

billwatersiii replied:
@kartben , I think that it has to remain where it is.  It applies to more that the pwm driver.  It also applies to our timer and counter drivers.

I will update the #include guard.

billwatersiii replied:
https://github.com/zephyrproject-rtos/zephyr/pull/95870/commits/a0a2b5df7166b9acfcbdd431a34f72fc2a1cbb93#diff-10aa3e5ac3b34e9dd53dba3ae64f0fd2fc2926d43670c29026e9149368f30459R22
